servers/src/grin/sync/header_sync.rs

```diff
@@ -117,6 +122,43 @@ impl HeaderSync {
 				header_head.height,
 				header_head.height,
 			);
+
+			// save the stalling start time
+			if stalling {
+				if self.stalling_ts.is_none() {
+					self.stalling_ts = Some(now);
+				}
+			} else {
+				self.stalling_ts = None;
+			}
+
+			if !all_headers_received {
+				if let Some(ref stalling_ts) = self.stalling_ts {
+					if let Some(ref peer) = self.syncing_peer {
+						match self.sync_state.status() {
+							SyncStatus::HeaderSync {
+								current_height: _,
+								highest_height,
+							} => {
+								// Ban this fraud peer which claims a higher work but can't send us the real headers
+								if now > *stalling_ts + Duration::seconds(120)
+									&& highest_height == peer.info.height()
```

Comment:
I still find myself struggling with Rust from time to time, but If I understand this code correctly, you're saving stalling_ts the first time we stall during header sync. Then, if 120 seconds pass, and we stall again (potentially with a different peer), we ban the peer that we most recently requested from? That doesn't seem like the right thing to do.

Reply:
Damn! You're right, I completely forgot to clear `self.stalling_ts` when syncing goes well.
thanks @DavidBurkett

Reply:
Updated for reset `self.stalling_ts` when syncing goes well.